src/ioctls/vcpu.rs

```
@@ -853,6 +853,41 @@ impl VcpuFd {
         Ok(xsave)
     }
 
+    /// X86 specific call that returns the vcpu's current "xsave struct" via `KVM_GET_XSAVE2`.
+    ///
+    /// See the documentation for `KVM_GET_XSAVE2` in the
+    /// [KVM API doc](https://www.kernel.org/doc/Documentation/virtual/kvm/api.txt).
+    ///
+    /// # Arguments
+    ///
+    /// * `vm_fd` - the file descriptor of the VM this [`Vcpu`] belongs to.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # extern crate kvm_ioctls;
+    /// # use kvm_ioctls::Kvm;
+    /// let kvm = Kvm::new().unwrap();
+    /// let vm = kvm.create_vm().unwrap();
+    /// let vcpu = vm.create_vcpu(0).unwrap();
+    /// let xsave = vcpu.get_xsave2(&vm).unwrap();
+    /// ```
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    pub fn get_xsave2(&self, vm_fd: &crate::VmFd) -> Result<Xsave> {
+        let xsave_size = vm_fd.xsave_size();
```

I'm not sure if we really should get the required size internally.

VMM may check KVM caps beforehand (to fail faster). In that case, VMM can cache the XSAVE size somewhere and reuse it without calling KVM_CHECK_EXTENSIONS. For more flexibility of the VMM implementation, it could be better to keep `xsave_size()` out of `get_xsave2()`.

Rather, not having `xsave_size()` in the first place might be better. In the first look, handling both cases (returning 4KB if XSAVE2 is not supported) looks nice. But a VMM supporting multiple kernel versions supporting KVM_GET_XSAVE and KVM_GET_XSAVE2 (like firecracker) should check the availability of the new API first, and then, if not available, it should fall back to KVM_GET_XSAVE. If we really want to absorbing the differences of these two APIs, we should implement the fallback part. Otherwise, we should return an error early when retrieving the required size.

---

Mhh, yeah, I think I see what you mean. In the context of `get_xsave2`, having the fallback of `xsave_size()` to be 4096 when CAP_XSAVE2 isnt supported it fairly useless, because the KVM_GET_XSAVE2 call is gonna fail anyway. So it's kinda half-baked.

The problem API wise is that we cannot let the user allocate the Xsave struct, because if they allocate a too-small one, then get_xsave2 will result in undefined behavior potentially (e.g. the kernel writing past the allocated buffer). We could mark it unsafe, but that's also not very nice imo.

And yeah, I guess if all Xsave structs are created from within kvm-ioctls, then its not even useful to expose xsave_size() as an API altogether.

---

At least for me, marking it unsafe isn't so bad. `unsafe` marked functions always make me (hopefully all Rustaceans) pay my highest attention to it. Of course, safe wrapper is the best, but as being discussed in another thread, it may be impossible to wrap it safely due to the possible race condition.

So my current humble suggestion is:
- Mark `get_xsave2()` unsafe
- Make `get_xsave2()` fail early if KVM_CAP_XSAVE2 isn't supported
- Add lower wrappers like we do for `check_extension()`
  - Add `get_xsave2_xsave()`  which allows users to pass `&mut Xsave`.
  - Add `get_xsave2_raw()` which allows users to pass a pointer to `struct kvm_xsave`.